### PR TITLE
Alternative IFF

### PR DESCRIPTION
**Pull request by ihatethisengine**
# About the pull request

IFF no longer allows you to shoot past allies, instead it just prevents you from shooting if there is a friendly on your line of fire. Bullets still cannot hurt friendlies if something goes wrong. 

Rebalanced some of the weapons to make them work with the change:
Smartgun: increased damage from 30 to 36, increased armor piercing bullet damage from 20 to 24.
Smartpistol: increased damage from 30 to 36.
M56D: increased damage from 35 to 36, reduced firedelay from 0.3 to 0.25 for full-auto.
Smart-scope: removed damage reduction, added additional 8 ticks firedelay when IFF procs, so accuracy still matters. Also halved the time it takes to activate the scope.
M46C: no longer becomes weaker when IFF is enabled, also can use full-auto and burst-fire with IFF.

The rest of the guns (including M42A, GL, flaregun) still use the old system.

Most of the work done by MorrowWolf, I ported it from his project: https://github.com/PvE-CMSS13/PvE-CMSS13.

# Explain why it's good for the game

IFF always has been an issue. Potentially it's very strong, because it can be stacked. But because of its potential, guns with IFF on their own are made very weak, which is only further encourage players to murderball and stack IFF. It's a nightmare for balance and for gameplay, so this is why I am trying to change it. Also it makes a bit more sense than magic bullets and this is more lore-accurate AFAIK. 

# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

https://github.com/cmss13-devs/cmss13/assets/115417687/ad599ca8-92e0-4a5a-aa6e-0ce3d7fd97c5

</details>

# Changelog
:cl: MorrowWolf, ihatethisengine
add: Added new IFF system, that prevents user from firing weapon if there are friendlies on the line of fire.
balance: Smartgun, smartpistol, M56D, M46C and smart-scope all use new IFF system now.
balance: Increased smartgun damage from 30 to 36, increased armor piercing bullet damage from 20 to 24.
balance: Increased smartpistol damage from 30 to 36.
balance: Slightly increased damage and firerate of M56D.
balance: Smart-scope no longer reduces damage, guns with smart-scope get firedelay when user tries to shoot a friendly. Also halved the time it takes to activate the scope.
balance: M46C doesn't become weaker when IFF is enabled.
/:cl:

### CODE REVIEW

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

File: code/datums/components/iff_fire_prevention.dm [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	checked_turfs -= starting_turf
+
+	//At some angles (scatter or otherwise) the original target is not in checked_turfs so we put it in there in order based on distance from user
+	//If we are literally clicking on someone with IFF then we don't want to fire, feels funny as a user otherwise
+	if(projectile_to_fire.original)
+		var/turf/original_target_turf = get_turf(projectile_to_fire.original)
+
+		if(original_target_turf && !(original_target_turf in checked_turfs))
+			var/user_to_target_dist = get_dist(starting_turf, original_target_turf)
+			var/list/temp_checked_turfs = checked_turfs
```
Comment (Zonespace27) [inline code/datums/components/iff_fire_prevention.dm:46]: This won't work. Doing this is just making a 2nd pointer to the list. Append .Copy() to checked_turfs

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: code/_onclick/hud/fullscreen.dm [hunk 2]
```diff
@@ -61,6 +61,13 @@
 			else
 				client.remove_from_screen(screen)
 
+//Get the distance to the farthest edge of the screen
+/mob/proc/get_maximum_view_range()
+	if(!client)
+		return 7
+
+	var/offset = client.pixel_x || client.pixel_y
```
[the remark this answers is not in the document]
Comment (ihatethisengine) [reply code/_onclick/hud/fullscreen.dm:69]: It does not but I don't think we get situations when we have both x and y offset? I can just change it tho.

[the remark this answers is not in the document]
Comment (Zonespace27) [reply code/_onclick/hud/fullscreen.dm]: We do, and please do so

[the remark this answers is not in the document]
Comment (ihatethisengine) [reply code/_onclick/hud/fullscreen.dm]: I did, should be better I think?

**Comment by Sylphily** [comment]
If i ask, why is MD change here, has nothing to due with IFF at all, and doesn't sound like there is a good reason for it

**Comment by ihatethisengine** [comment]
That's simple. People are going to focus this change, so when we revert it, they won't have energy to fight the rest.

On a serious note, I don't mind dropping the change, but maybe whoever in charge of balance feels the same about MD.

**Comment by Sylphily** [comment]
yeah the MD thing seems like it came out of left field, and after asking morrow it seems like a thing to change how the SG gets a MD, it is instead they battle buddy role gets the MD and tells them where to fire, so it kinda split up here from what they intended!

File: code/modules/projectiles/guns/rifles.dm [hunk 3]
```diff
@@ -630,19 +632,10 @@
...
-	. = ..()
-	if(iff_enabled)
-		modify_fire_delay(FIRE_DELAY_TIER_12)
-		remove_firemode(GUN_FIREMODE_BURSTFIRE)
-		remove_firemode(GUN_FIREMODE_AUTOMATIC)
-
-	else
-		add_firemode(GUN_FIREMODE_BURSTFIRE)
-		add_firemode(GUN_FIREMODE_AUTOMATIC)
+		GetExactComponent(/datum/component/iff_fire_prevention).RemoveComponent()
```
Comment (fira) [inline code/modules/projectiles/guns/rifles.dm:638]: i think you should rather have the component "pause" with a simple variable, you're not really supposed to use GetComponent

**Review by fira — CHANGES_REQUESTED** [review]

File: code/__DEFINES/dcs/signals/atom/signals_gun.dm [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+/// from /obj/item/weapon/gun/proc/recalculate_attachment_bonuses() : ()
+#define COMSIG_GUN_RECALCULATE_ATTACHMENT_BONUSES "gun_recalculate_attachment_bonuses"
+
+/// from  /obj/item/weapon/gun/proc/load_into_chamber() : ()
+#define COMSIG_GUN_INTERRUPT_FIRE "gun_interrupt_fire"
+
+//Signals for automatic fire at component
+#define COMSIG_AUTOMATIC_SHOOTER_START_SHOOTING_AT "start_shooting_at"
+#define COMSIG_AUTOMATIC_SHOOTER_STOP_SHOOTING_AT "stop_shooting_at"
+#define COMSIG_AUTOMATIC_SHOOTER_SHOOT "shoot"
```
[the remark this answers is not in the document]
Comment (ihatethisengine) [reply code/__DEFINES/dcs/signals/atom/signals_gun.dm:22]: uh it just was moved from signals_item here, do you have suggestions how to rename these?

File: code/modules/projectiles/guns/rifles.dm (hunk 3 above)
Comment (ihatethisengine) [reply code/modules/projectiles/guns/rifles.dm:638]: And how do I pass a variable to a component without getting it? Can you give an example?

Comment (ihatethisengine) [reply code/modules/projectiles/guns/rifles.dm:638]: Probably I can make a new signal but eh...

File: code/__DEFINES/dcs/signals/atom/signals_gun.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (Zonespace27) [reply code/__DEFINES/dcs/signals/atom/signals_gun.dm:22]: the convention for signal names is `#define COMISG_FOO_BAR_AEIOU "foo_bar_aeiou"

[the remark this answers is not in the document]
Comment (ihatethisengine) [reply code/__DEFINES/dcs/signals/atom/signals_gun.dm:22]: It doesn't even seem to be used but oh well

File: code/modules/projectiles/gun.dm [hunk 5]
```diff
@@ -1670,26 +1676,40 @@ not all weapons use normal magazines etc. load_into_chamber() itself is designed
...
+	var/actual_sound = fire_sound
+	if(isnull(fire_sound))
+		actual_sound = pick(fire_sounds)
+
+	if(projectile_to_fire.ammo && projectile_to_fire.ammo.sound_override)
+		actual_sound = projectile_to_fire.ammo.sound_override
+
+	projectile_to_fire.firer = user
+	if(isliving(user))
+		projectile_to_fire.def_zone = user.zone_selected
```
Comment (fira) [inline code/modules/projectiles/gun.dm:1695]: wait hold on this doesn't belong in `play_firing_sounds`

**Review by fira — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/gun.dm (hunk 5 above)
Comment (ihatethisengine) [reply code/modules/projectiles/gun.dm]: Is this better?

**Comment by stagda** [comment]
When does the check occur as to whether or not to fire? It seems like its checking to see if the bullet hits first? I've noticed that in a quick check it will still fire if someone is in the way if it calculates the bullet wont hit. E.g. I aim at someone in front of me, but since the bullet will hit the wall next to them due to scatter it will fire despite the point of aim having them in the way.   

Could be a bit confusing in heat of the moment to have it seem like sometimes you are blocked and others you aren't

https://github.com/cmss13-devs/cmss13/assets/106484997/454e83d7-f0b3-449d-8efb-628c66a5bf24

**Comment by BroseidonMOO** [comment]
Testmerged during round 19159, no stability or bug related issues noticed. Balance wise, it is worth noting the attrition rate was pretty high for SGs. There were multiple lag complaints, but unclear if that was related to PR.

**Comment by ihatethisengine** [comment]
> When does the check occur as to whether or not to fire?

Yeah it checks where bullet is going to hit to take recoil & scatter into account, instead of just checking wherever you are clicking.  Although there is an issue that canceled shot causes scatter anyway, need to look into this.

**Comment by Wintermote** [comment]
Smartscopes also reduce the rate of fire of guns they're on, not just the damage

Also you are able to PB marines with the new smartscope equipped, I'm not sure if that's intended or not.

**Comment by ihatethisengine** [comment]
> Smartscopes also reduce the rate of fire of guns they're on, not just the damage
Yeah as they used to. All scopes except X2 reduce firerate so I thought it would be a fair trade-off, because there probably should be one.

> Also you are able to PB marines with the new smartscope equipped, I'm not sure if that's intended or not.
Apparently you always were able to, because the game counts PB as a melee attack. Doubt that was ever intended, so I'll look into it.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

**Pull request closed without merging**